lib/mnesia/src/mnesia.erl

```diff
@@ -838,18 +838,20 @@ read(Tid, Ts, Tab, Key, LockKind)
 	tid ->
 	    Store = Ts#tidstore.store,
 	    Oid = {Tab, Key},
-	    Objs =
-		case LockKind of
-		    read ->
-			mnesia_locker:rlock(Tid, Store, Oid);
-		    write ->
-			mnesia_locker:rwlock(Tid, Store, Oid);
-		    sticky_write ->
-			mnesia_locker:sticky_rwlock(Tid, Store, Oid);
-		    _ ->
-			abort({bad_type, Tab, LockKind})
-		end,
-	    add_written(?ets_lookup(Store, Oid), Tab, Objs);
+	    ObjsFun =
+                fun() ->
+                        case LockKind of
+                            read ->
+                                mnesia_locker:rlock(Tid, Store, Oid);
+                            write ->
+                                mnesia_locker:rwlock(Tid, Store, Oid);
+                            sticky_write ->
+                                mnesia_locker:sticky_rwlock(Tid, Store, Oid);
+                            _ ->
+                                abort({bad_type, Tab, LockKind})
+                        end
+                end,
```

Comment:
Yes, both of the risks were considered while creating this fix.

An explicit check for lock types can be added here or it would have been nice if mnesia:lock/2 could be called instead if written data already exists in transaction context.
However, mnesia:lock/2 uses the same function, ie mnesia_locker:rlock/3, and thus reads the record. And even though mnesia:lock/2 is documented to return 'ok' for read locks it actually returns the object read (and mnesia:lock doesnt't support sticky_write either).

It was also noted that sticky_write locks are not considered if a write lock already exists and thus it was seen safe not to evaluate ObjsFun here.
 
@dgud Do you think this comment needs to be fixed?

Reply:
The verification of correct lock type would be easy to implement:
```erlang
LockFun =
    case LockKind of
        read -> fun mnesia_locker:rlock/3;
        write -> fun mnesia_locker:rwlock/3;
        sticky_write -> fun mnesia_locker:sticky_rwlock/3;
        _ -> abort({bad_type, Tab, LockKind})
    end,
ObjsFun = fun () -> LockFun(Tid, Store, Oid) end,
...
```

My second concern would be enough to be documented in a comment I guess.